### PR TITLE
2508 add loki source kafka component

### LINKED ISSUE #2508: `loki.source.kafka`: Reads messages from Kafka using a consumer group

### PR DESCRIPTION
**Pull request by erikbaranowski**
#### PR Description

This PR pulls over promtail code from loki for kafka and wires it up as a flow component `loki.source.kafka`

#### Which issue(s) this PR fixes

Closes #2508 

#### PR Checklist

- [x] CHANGELOG updated
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

File: docs/sources/flow/reference/components/loki.source.kafka.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+}
+```
+
+## Arguments
+
+`loki.source.kafka` supports the following arguments:
+
+Name                     | Type                   | Description          | Default | Required
+------------------------ | ---------------------- | -------------------- | ------- | --------
+`brokers`                | `list(string)`         | The list of brokers to connect to kafka.                 |              | yes
```
Comment (karengermond) [inline docs/sources/flow/reference/components/loki.source.kafka.md:37]:
```suggestion
`brokers`                | `list(string)`         | The list of brokers to connect to Kafka.                 |              | yes
```

File: docs/sources/flow/reference/components/loki.source.kafka.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+`topics`                 | `list(string)`         | The list of Kafka topics to consume.                     |              | yes
+`group_id`               | `string`               | The Kafka consumer group id.                             | `"promtail"` | no
+`assignor`               | `string`               | The consumer group rebalancing strategy to use.          | `"range"`    | no
+`version`                | `string`               | Kafka version to connect to.                             | `"2.2.1"`    | no
+`use_incoming_timestamp` | `bool`                 | Whether or not to use the timestamp received from Kafka. | `false`      | no
+`labels`                 | `map(string)`          | The labels to associate with each received Kafka event.  | `{}`         | no
+`forward_to`             | `list(LogsReceiver)`   | List of receivers to send log entries to.                |              | yes
+`relabel_rules`          | `RelabelRules`         | Relabeling rules to apply on log entries.                | `{}`         | no
+
+`assignor` values can be either `"range"`, `"roundrobin"` or `"sticky"`.
```
Comment (karengermond) [inline docs/sources/flow/reference/components/loki.source.kafka.md:47]:
```suggestion
`assignor` values can be either `"range"`, `"roundrobin"`, or `"sticky"`.
```

File: docs/sources/flow/reference/components/loki.source.kafka.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+authentication > sasl_config | [sasl_config] | Optional authentication configuration with Kafka brokers. | no
+authentication > sasl_config > tls_config | [tls_config] | Optional authentication configuration with Kafka brokers. | no
+
+[authentication]: #authentication-block
+[tls_config]: #tls_config-block
+[sasl_config]: #sasl_config-block
+
+### authentication block
+
+The `authentication` block defines the authentication method when communicating with the kafka event brokers.
```
Comment (karengermond) [inline docs/sources/flow/reference/components/loki.source.kafka.md:70]:
```suggestion
The `authentication` block defines the authentication method when communicating with the Kafka event brokers.
```

File: docs/sources/flow/reference/components/loki.source.kafka.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+Name                     | Type          | Description | Default | Required
+------------------------ | ------------- | ----------- | ------- | --------
+`type`                   | `string`      | Type is authentication type. | `"none"` | no
```
Comment (karengermond) [inline docs/sources/flow/reference/components/loki.source.kafka.md:74]:
```suggestion
`type`                   | `string`      | Type of authentication. | `"none"` | no
```

File: docs/sources/flow/reference/components/loki.source.kafka.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+`mechanism` | `string` | Specifies the SASL mechanism the client uses to authenticate with the broker. | `"PLAIN""` | no
+`user`      | `string` | The user name to use for SASL authentication. | | no
+`password`  | `string` | The password to use for SASL authentication. | | no
+`use_tls`   | `bool`   | If true, SASL authentication is executed over TLS. | `false` | no
+
+## Labels
+
+The `labels` map is applied to every message that the component reads.
+
+The following internal labels all prefixed with `__` are available but will be discarded if not relabeled:
```
Comment (karengermond) [inline docs/sources/flow/reference/components/loki.source.kafka.md:99]:
```suggestion
The following internal labels prefixed with `__` are available but are discarded if not relabeled:
```

**Review by karengermond — APPROVED** [review]
LGTM! A few comments...

File: docs/sources/flow/reference/components/loki.source.kafka.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
+---
+title: loki.source.kafka
+---
+
+# loki.source.kafka
+
+`loki.source.kafka` reads messages from Kafka using a consumer group
+and forwards them to other `loki.*` components.
+
+The component starts a new kafka consumer group for the given arguments
```
Comment (rfratto) [inline docs/sources/flow/reference/components/loki.source.kafka.md:10]:
```suggestion
The component starts a new Kafka consumer group for the given arguments
```

File: docs/sources/flow/reference/components/loki.source.kafka.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+Multiple `loki.source.kafka` components can be specified by giving them
+different labels.
+
+## Usage
+
+```river
+loki.source.kafka "LABEL" {
+	brokers    = BROKER_LIST
+	topics     = TOPIC_LIST
+    forward_to = RECEIVER_LIST
```
Comment (rfratto) [inline docs/sources/flow/reference/components/loki.source.kafka.md:27]:
```suggestion
	brokers    = BROKER_LIST
	topics     = TOPIC_LIST
	forward_to = RECEIVER_LIST
```

**Review by rfratto — APPROVED** [review]
LGTM, but I'd like for @tpaschalis to take a look since he's more familiar with these components.

File: component/loki/source/kafka/kafka.go [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		c.mut.Lock()
+		defer c.mut.Unlock()
+
+		level.Info(c.opts.Logger).Log("msg", "loki.source.kafka component shutting down, stopping listener")
+		if c.target != nil {
+			err := c.target.Stop()
+			if err != nil {
+				level.Error(c.opts.Logger).Log("msg", "error while stopping kafka listener", "err", err)
```
Comment (tpaschalis) [inline component/loki/source/kafka/kafka.go:132]: nit: s/listener/target/

File: docs/sources/flow/reference/components/loki.source.kafka.md [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+`mechanism` | `string` | Specifies the SASL mechanism the client uses to authenticate with the broker. | `"PLAIN""` | no
+`user`      | `string` | The user name to use for SASL authentication. | | no
+`password`  | `string` | The password to use for SASL authentication. | | no
+`use_tls`   | `bool`   | If true, SASL authentication is executed over TLS. | `false` | no
+
+## Labels
+
+The `labels` map is applied to every message that the component reads.
+
+The following internal labels prefixed with `__` are available but are discarded if not relabeled:
```
Comment (tpaschalis) [inline docs/sources/flow/reference/components/loki.source.kafka.md:99]: Question: Is this component susceptible to the same issue as other components where no labels are populated by default, therefore by discarding these we'll get an empty label set and the message will never be sent to Loki?

File: docs/sources/flow/reference/components/loki.source.kafka.md [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+### sasl_config block
+
+The `sasl_config` block defines the listen address and port where the listener
+expects Kafka messages to be sent to.
+
+Name                     | Type          | Description | Default | Required
+------------------------ | ------------- | ----------- | ------- | --------
+`mechanism` | `string` | Specifies the SASL mechanism the client uses to authenticate with the broker. | `"PLAIN""` | no
+`user`      | `string` | The user name to use for SASL authentication. | | no
+`password`  | `string` | The password to use for SASL authentication. | | no
```
Comment (tpaschalis) [inline docs/sources/flow/reference/components/loki.source.kafka.md:92]:
```suggestion
`user`      | `string` | The user name to use for SASL authentication. | `""` | no
`password`  | `string` | The password to use for SASL authentication. | `""` | no
```

**Review by tpaschalis — APPROVED** [review]
LGTM apart for a few nits, nice work! (I only skimmed over the ported Promtail code)

File: docs/sources/flow/reference/components/loki.source.kafka.md [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+
+## Arguments
+
+`loki.source.kafka` supports the following arguments:
+
+Name                     | Type                   | Description          | Default | Required
+------------------------ | ---------------------- | -------------------- | ------- | --------
+`brokers`                | `list(string)`         | The list of brokers to connect to Kafka.                 |              | yes
+`topics`                 | `list(string)`         | The list of Kafka topics to consume.                     |              | yes
+`group_id`               | `string`               | The Kafka consumer group id.                             | `"promtail"` | no
```
Comment (tpaschalis) [inline docs/sources/flow/reference/components/loki.source.kafka.md:39]: Let's also change the default here

File: docs/sources/flow/reference/components/loki.source.kafka.md (hunk 9 above)
Comment (erikbaranowski) [reply docs/sources/flow/reference/components/loki.source.kafka.md:99]: that sounds right to me, but I don't know 100% how it works

Comment (erikbaranowski) [reply docs/sources/flow/reference/components/loki.source.kafka.md:99]: I tested and yes, this is how it works

File: component/loki/source/kafka/kafka.go [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+		return err
+	}
+
+	c.target = t
+
+	return nil
+}
+
+// Convert is used to bridge between the River and Promtail types.
+func (args *Arguments) Convert() scrapeconfig.Config {
```
[the remark this answers is not in the document]
Comment (erikbaranowski) [reply component/loki/source/kafka/kafka.go:152]: @rfratto might take a crack at it considering the time constraint either from my in-progress scrapeconfig commit or a fresh implementation (we can revert that last commit if needed). big thanks in advance on the collaboration.

[the remark this answers is not in the document]
Comment (rfratto) [reply component/loki/source/kafka/kafka.go]: I literally copied the Promtail types into the internal package. If that should be changed to use the Flow arguments type throughout, let me know :)

**Pull request merged**